### PR TITLE
Refactoring SolidModelConverter.h

### LINKED ISSUE #320: [REF] SolidConverter
**File which is being refactored**
core/IfcGeometryConverter/SolidConverter

**Functions**
all functions except for `IfcSectionedSolid` - see #156 

**Instruction**
@SamuilsRulovs : if you see fit, you can assign individual functions to @Elvira2227 - please produce a separate issue for that.

### LINKED ISSUE #345: [BUG] SectionedSolidHorizontal
SectionedSolidHorizontal is displayed incorrectly.
**Describe the bug**
Calculations for SectionedSolidHorizontal example file are incorrect, thus SectionedSolidHorizontal is displayed incorrectly

**To Reproduce**
1. Select OpenIfraPlatform.UI as StartUp Project
2. Load file “SSH-Curve_1.ifc”
3. See error

**Expected behavior**
SectionedSolidHorizontal should be displayed correctly. 

**Screen shot**
Screenshot 2021-01-22 173526

**CMake configuration**
If applicable, list any special configurations you have selected in CMake.
- EARLYBINDING_WITH_IFC4x1: selected

**Additional context**

#156 - (HINT: it might be something wrong with the angle value)

### LINKED ISSUE #356: [REF] SolidModelConverter.h
**File which is being refactored**
SolidModelConverter.h

**Functions**
List all functions in the file which have to be refactored with possible code snippets.

- [ ] convertIfcSolidModel
- [ ] convertIfcCsgSolid
- [ ] convertIfcManifoldSolidBrep
- [ ] convertIfcSectionedSolidHorizontal

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Fixes #356 

Fixes #320

### CODE REVIEW

**Review by pjanck — COMMENTED** [review]
Some guidelines:
- use `const noexcept(false)` for function signatures
- all function should follow the same order of input parameters - reference to the entity, pos, itemData (it's easier to follow the code)

**Comment by SamuilsRulovs** [comment]
Excuse me for a little chaos, I am only staring to understand connections between entities in SolidModelConverter.h.

**Review by pjanck — COMMENTED** [review]
The diff is pure chaos. Thank god you've already prepared many unit tests so that none of the functionalities can go missing :) Commenting, so that I can focus on individual changes later. I trust that your changes can compile and are fine.

**Review by ibrahem36192012 — APPROVED** [review]

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 1]
```diff
@@ -860,103 +862,117 @@ namespace OpenInfraPlatform
...
+					throw oip::InconsistentModellingException(extrudedArea, "Invalid Depth!");
 				}
-				double length_factor = UnitConvert()->getLengthInMeterFactor();
 
 				// direction and length of extrusion
-				const double depth = (typename IfcEntityTypesT::IfcLengthMeasure)(extrudedArea->Depth) * length_factor;
-				//const double depth = extrudedArea->Depth->length_factor;
-				carve::geom::vector<3>  extrusion_vector;
-				auto& vec_direction = extrudedArea->ExtrudedDirection->DirectionRatios;
+				const double depth = (typename IfcEntityTypesT::IfcLengthMeasure)(extrudedArea->Depth) * UnitConvert()->getLengthInMeterFactor();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: When calling `UnitConvert()` use `this->UnitConvert()` - it's an error under VS2019

Also, no need for casting to `IfcLengthMeasure`.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 2]
```diff
@@ -860,103 +862,117 @@ namespace OpenInfraPlatform
...
 				{
-					extrusion_vector = carve::geom::VECTOR(vec_direction[0] * depth, vec_direction[1] * depth, 0);
+					extrusionVector.x = vecDirection[0] * depth;
+					extrusionVector.y = vecDirection[1] * depth;
+					break;
 				}
-
+				default:
+					throw oip::InconsistentModellingException(extrudedArea, "Cardinality of the direction ratios can be only 2 or 3");
+				}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Use placement converter's `convertIfcDirection`

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 3]
```diff
@@ -1722,21 +1730,21 @@ namespace OpenInfraPlatform
...
 				if (block)
 				{
-					double x_length = length_factor;
-					double y_length = length_factor;
-					double z_length = length_factor;
+					double x_length = UnitConvert()->getLengthInMeterFactor();
+					double y_length = UnitConvert()->getLengthInMeterFactor();
+					double z_length = UnitConvert()->getLengthInMeterFactor();
 
 					if ((typename IfcEntityTypesT::IfcLengthMeasure)block->XLength)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: I believe these `->XLength`... are not optional - so no need to check for their existence.

**Review by pjanck — COMMENTED** [review]
Apply the comments appropriately everywhere.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 4]
```diff
@@ -419,17 +419,17 @@ namespace OpenInfraPlatform
...
+					const std::vector<std::vector<carve::geom::vector<2> > >& profileCoords = profileConverter->getCoordinates();
 
 					// Save profile coords in paths
-					std::vector<std::vector<carve::geom::vector<2> > > profile_coords_2d;
-					for (int p = 0; p < profile_coords.size(); ++p)
+					std::vector<std::vector<carve::geom::vector<2> > > profileCoords2D;
+					for (int p = 0; p < profileCoords.size(); ++p)
 					{
-						const std::vector<carve::geom::vector<2> >& profile_loop = profile_coords[p];
+						const std::vector<carve::geom::vector<2> >& profileLoop = profileCoords[p];
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: this could be simplified with

`for( const auto& profileLoop : profileCoords )`

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 5]
```diff
@@ -945,25 +945,8 @@ namespace OpenInfraPlatform
...
-				case 2:
-				{
-					extrusionVector.x = vecDirection[0] * depth;
-					extrusionVector.y = vecDirection[1] * depth;
-					break;
-				}
-				default:
-					throw oip::InconsistentModellingException(extrudedArea, "Cardinality of the direction ratios can be only 2 or 3");
-				}
+				const double depth = (typename IfcEntityTypesT::IfcLengthMeasure)(extrudedArea->Depth) * this->UnitConvert()->getLengthInMeterFactor();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]:
```suggestion
				const double depth = extrudedArea->Depth * this->UnitConvert()->getLengthInMeterFactor();
```

No need for casts.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 6]
```diff
@@ -976,7 +959,7 @@ namespace OpenInfraPlatform
 
 				if (paths.size() == 0)
 				{
-					return;
+					throw oip::InconsistentModellingException(extrudedArea, "Paths are empty!");
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: That would be a geometry exception.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 7]
```diff
@@ -1014,40 +997,31 @@ namespace OpenInfraPlatform
...
-					if (axis_placement->Axis)
+					if (axisPlacement->Axis)
 					{
-						decltype(axis_placement->Axis)::type axis = axis_placement->Axis;
-						axis_direction = carve::geom::VECTOR(
+						decltype(axisPlacement->Axis)::type axis = axisPlacement->Axis;
+						axisDirection = carve::geom::VECTOR(
 							axis->DirectionRatios[0],
 							axis->DirectionRatios[1],
 							axis->DirectionRatios[2]);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: use placement converter's `convertIfcDirection`

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 8]
```diff
@@ -1056,195 +1030,158 @@ namespace OpenInfraPlatform
...
-					carve::geom2d::P2& point_projected = loop_projected[index_in_loop];
-					merged.push_back( point_projected );
-					}
-					}
-					carve::triangulate::triangulate(merged, triangulated);
-					carve::triangulate::improve(merged, triangulated);*/
-
-
 					BLUE_LOG(error) << "carve::triangulate::incorporateHolesIntoPolygon failed ";
 					return;
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Either/or:

- Consider introducing a new exception
- throw a geometry exception

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 9]
```diff
@@ -2087,7 +2024,7 @@ namespace OpenInfraPlatform
 				if (otherOperand)
 				{
 					carve::geom::aabb<3> aabb;
-					otherOperand->createMeshSetsFromClosedPolyhedrons();
+					otherOperand->createMeshSetsFromclosed_polyhedrons();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: This is a weird function name ...

**Review by pjanck — CHANGES_REQUESTED** [review]

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 10]
```diff
@@ -1015,7 +1026,7 @@ namespace OpenInfraPlatform
 
 					if (axisPlacement->Location)
 					{
-						axisLocation = placementConverter->convertIfcCartesianPoint(axisPlacement->Location );
+						axisLocation = placementConverter->convertIfcPoint(axisPlacement->Location);
 					}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: I assume this can be greatly simplified by using specialized functions in PlacementConverter.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 9 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: I changed it by mistake. Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 8 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Right now I've decided to stay with InconsistentGeometryException, but I will try to introduce a new exception after I am done with this. Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 7 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 6 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 5 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 4 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: I have't come so far (I was changing only length_factor to  UnitConvert()->getLengthInMeterFactor() in whole file). I am almost there, so I will correct it as soon as I am there.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 2 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 1 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4c40ba7

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 3 above)
Comment (pjanck) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Right, understood!

File: Core/src/IfcGeometryConverter/GeomUtils.cpp [hunk 11]
```diff
@@ -1043,6 +1043,29 @@ void GeomUtils::convertPlane2Matrix( const carve::geom::vector<3>& plane_normal,
...
+	const carve::math::Matrix & positionMatrix,
+	const int ammountOfPoints,
+	std::vector<carve::geom::vector<3>>& newListOfPoints
+) noexcept(false) {
+	for (int i = 0; i < ammountOfPoints; ++i)
+	{
+		carve::geom::vector<3> vertex = listOfPoints.at(i);
+		vertex = positionMatrix * vertex;
+		newListOfPoints.push_back(vertex);
+	}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/GeomUtils.cpp]: Simpler code:

```suggestion
	for ( const auto& vertex : listOfPoints )
	{
		newListOfPoints.push_back(positionMatrix * vertex);
	}
```

Therefore, no need for the amount of points being passed.

Doxycommnt names of variables do not match.

Comment (pjanck) [reply Core/src/IfcGeometryConverter/GeomUtils.cpp]: Additionally: if the size of the array is known upfront, it is advisable to reserve the amount of space in advance (speed optimisation). Prepend the for loop with

    newListOfPoints.reserve( listOfPoints.size() );

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 12]
```diff
@@ -855,7 +853,8 @@ namespace OpenInfraPlatform
 				throw oip::UnhandledException(sweptAreaSolid);
 			}
 
-			void convertIfcFixedReferenceSweptAreaSolid(EXPRESSReference<typename IfcEntityTypesT::IfcFixedReferenceSweptAreaSolid> fixedRefSweptAreaSolid,
+			void convertIfcFixedReferenceSweptAreaSolid(
+				EXPRESSReference<typename IfcEntityTypesT::IfcFixedReferenceSweptAreaSolid> fixedRefSweptAreaSolid,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: shouldn't this be a `const &` reference?

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 13]
```diff
@@ -1196,268 +1201,295 @@ namespace OpenInfraPlatform
...
-					// TODO: implement and check for formal propositions.
+				// IfcSweptDiskSolidPolygonal SUBTYPE of IfcSweptDiskSolid
+				if (sweptDiskSolid.template isOfType<typename IfcEntityTypesT::IfcSweptDiskSolidPolygonal>())
+				{
+					convertIfcSweptDiskSolidPolygonal(sweptDiskSolid.template as<typename IfcEntityTypesT::IfcSweptDiskSolidPolygonal>(),
+						pos, itemData);
+					//TODO: implement and check for formal propositions.
 
-				//} //endif swept_disk_solid_polygonal
+				} //endif sweptDiskSolid_polygonal
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Shouldn't there be a return here?

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 14]
```diff
@@ -1196,268 +1201,295 @@ namespace OpenInfraPlatform
...
 				}
 
-				if (!bend_found)
+				if (!bendFound)
 				{
 					// sweeping curve is linear. assume any local z vector
 					local_z = carve::geom::VECTOR(0, 0, 1);
-					double dot_normal_local_z = dot((basis_curve_points.at(1) - basis_curve_points.at(0)), local_z);
+					double dot_normal_local_z = dot((basisCurvePoints.at(1) - basisCurvePoints.at(0)), local_z);
 					if (abs(dot_normal_local_z) < 0.0001)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: use `AreEqual()`

**Review by pjanck — COMMENTED** [review]
Slowly but surely taking a good form!

I like how you've created the new function `createFacesOfShape` (or sth like that) - this is used many times throughout the code.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 14 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 3f70103

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 13 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 3f70103

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 12 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 3f70103

File: Core/src/IfcGeometryConverter/GeomUtils.cpp (hunk 11 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/GeomUtils.cpp]: Resolved: 3f70103

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 15]
```diff
@@ -1326,7 +1327,7 @@ namespace OpenInfraPlatform
 					// sweeping curve is linear. assume any local z vector
 					local_z = carve::geom::VECTOR(0, 0, 1);
 					double dot_normal_local_z = dot((basisCurvePoints.at(1) - basisCurvePoints.at(0)), local_z);
-					if (abs(dot_normal_local_z) < 0.0001)
+					if (this->GeomSettings()->areEqual(abs(dot_normal_local_z), 0))
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]:
```suggestion
					if (this->GeomSettings()->areEqual(abs(dot_normal_local_z), 0., 0.0001))
```

Always use doubles (0.) and stick with the original comparison delta.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 16]
```diff
@@ -1625,6 +1600,30 @@ namespace OpenInfraPlatform
 				std::copy(firstOperandData->meshsets.begin(), firstOperandData->meshsets.end(), std::back_inserter(itemData->meshsets));
 
 			}
+			
+			/*! \brief Returns an ID of the selected geometric representations. 
+				*
+				* \param[in] ifcOperand				A list of the IDs for geometric representations.
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: this is not a list - it is a SELECT.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 17]
```diff
@@ -1625,6 +1600,30 @@ namespace OpenInfraPlatform
...
+				case 0:
+					return ifcOperand.get<0>().lock()->getId();
+				case 1:
+					return ifcOperand.get<1>().lock()->getId();
+				case 2:
+					return ifcOperand.get<2>().lock()->getId();
+				case 3:
+					return ifcOperand.get<3>().lock()->getId();
+				default:
+					return 0;
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: rather make a throw with notsupportedyet exception

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 18]
```diff
@@ -1625,6 +1600,30 @@ namespace OpenInfraPlatform
...
+				}
+			}
+
 
 			void convertIfcBooleanClippingResult(
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: IfcBooleanClippingResult is already supported with IfcBooleanResult - it is only a where rule that differentiates these two.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 49fa394

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 15 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved:  4bc3608

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 16 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4bc3608

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 17 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 4bc3608

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 18 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 0cba750

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 19]
```diff
@@ -1696,275 +1680,381 @@ namespace OpenInfraPlatform
...
-					polyhedron_data->addFace(5, 4, 7);
+				carve::math::Matrix primitivePlacementMatrix = block->Position ?
+					pos * placementConverter->convertIfcAxis2Placement3D(block->Position) :
+					pos;
 
-					polyhedron_data->addFace(0, 4, 5);
-					polyhedron_data->addFace(5, 1, 0);
+				double x_length = block->XLength * UnitConvert()->getLengthInMeterFactor();
+				double y_length = block->YLength * UnitConvert()->getLengthInMeterFactor();
+				double z_length = block->ZLength * UnitConvert()->getLengthInMeterFactor();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: should be `this->UnitConvert()...`

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 20]
```diff
@@ -1696,275 +1680,381 @@ namespace OpenInfraPlatform
...
+				double  y_length = rectangularPyramid->YLength * 0.5 * UnitConvert()->getLengthInMeterFactor();
+				double height = rectangularPyramid->Height * 0.5 * UnitConvert()->getLengthInMeterFactor();
 
-					int numVerticesInCircle = GeomSettings()->getNumberOfVerticesForTessellation(radius);
-					double deltaAngle = GeomSettings()->getAngleLength(radius);
-					for (double angle = 0.; angle < 2 * M_PI; angle += deltaAngle)
-					{
-						polyhedron_data->addVertex(primitive_placement_matrix*carve::geom::VECTOR(sin(angle)*radius, cos(angle)*radius, 0.0));
-					}
+				polyhedronData->addVertex(primitivePlacementMatrix*carve::geom::VECTOR(0, 0, height));
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: use 0. for values here

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 21]
```diff
@@ -1696,275 +1680,381 @@ namespace OpenInfraPlatform
...
+					pos;
+
+				double height = rightCircularCone->Height * UnitConvert()->getLengthInMeterFactor();
+				double radius = rightCircularCone->BottomRadius * UnitConvert()->getLengthInMeterFactor();
+
+				polyhedronData->addVertex(primitivePlacementMatrix*carve::geom::VECTOR(0.0, 0.0, height)); // top
+				polyhedronData->addVertex(primitivePlacementMatrix*carve::geom::VECTOR(0.0, 0.0, 0.0)); // bottom center
+
+				int numVerticesInCircle = GeomSettings()->getNumberOfVerticesForTessellation(radius);
+				double deltaAngle = GeomSettings()->getAngleLength(radius);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: use `this->`

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 22]
```diff
@@ -1696,275 +1680,381 @@ namespace OpenInfraPlatform
...
-					// uppper triangle fan
-					for (int i = 0; i < nvc - 1; ++i)
-					{
-						polyhedron_data->addFace(0, i + 2, i + 1);
-					}
-					polyhedron_data->addFace(0, 1, nvc);
+				const int numberOfVertices = GeomSettings()->getNumberOfSegmentsForTessellation(radius);
+				const double horizontalAngleDelta = GeomSettings()->getAngleLength(radius);
+				const int num_vertical_edges = ceil(0.5 * numberOfVertices);
+				double verticalAngleDelta = M_PI / double(num_vertical_edges - 1);	// TODO: adapt to model size and complexity
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: this todo is done, by using ->getNumberOfSegmentsForTessellation - the note can be removed

**Review by pjanck — CHANGES_REQUESTED** [review]
Apply the comments for all functions.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 22 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 207de31

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 21 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 207de31

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 20 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 207de31

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 19 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 207de31

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 23]
```diff
@@ -2058,326 +2058,364 @@ namespace OpenInfraPlatform
...
+				EXPRESSReference<typename IfcEntityTypesT::IfcSurface> base_surface = halfSpaceSolid->BaseSurface;
 
 				// base surface
-				std::shared_ptr<typename IfcEntityTypesT::IfcElementarySurface> elem_base_surface =
-					std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcElementarySurface>(base_surface);
-				if (!elem_base_surface)
-				{
-					BLUE_LOG(warning) << "The base surface shall be an unbounded surface (subtype of IfcElementarySurface)";
-					return;
-				}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: This should not be handled as currently proposed - it is only a check of the attribute, that it is of correct type.

Should be:

```
if( ! base_surface.template isOfType<typename IfcEntityTypesT::IfcElementarySurface>() )
    throw ...
```

With throwing inconsistent modelling exception...

**Review by pjanck — COMMENTED** [review]
It's hard to check the code, since the indentation changed ...

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 24]
```diff
@@ -461,32 +462,36 @@ namespace OpenInfraPlatform
...
-					if (fixed_axis_vertical == true)
+					if (fixedAxisVertical == true)
 					{
 						directionsOfCurve[pos].z = 0;
 						directionsOfCurve[pos].normalize();
 					}
+					else
+					{
+
+					}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Why is this needed? I believe this can be safely removed. There is no else.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 25]
```diff
@@ -948,8 +953,25 @@ namespace OpenInfraPlatform
 				}
 
 				// direction and length of extrusion
-				const double depth = extrudedArea->Depth * this->UnitConvert()->getLengthInMeterFactor();
-				carve::geom::vector<3> extrusionVector = placementConverter->convertIfcDirection(extrudedArea->ExtrudedDirection);
+				const double depth = (typename IfcEntityTypesT::IfcLengthMeasure)(extrudedArea->Depth) * UnitConvert()->getLengthInMeterFactor();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Is the cast to `IfcLengthMeasure` really needed?

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 26]
```diff
@@ -948,8 +953,25 @@ namespace OpenInfraPlatform
...
+				}
+				case 2:
+				{
+					extrusionVector.x = vecDirection[0] * depth;
+					extrusionVector.y = vecDirection[1] * depth;
+					break;
+				}
+				default:
+					throw oip::InconsistentModellingException(extrudedArea, "Cardinality of the direction ratios can be only 2 or 3");
+				}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: Please use placementconverter::convertIfcDirection

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 27]
```diff
@@ -1008,196 +1030,237 @@ namespace OpenInfraPlatform
...
+					}
+
+					if (axis_placement->Axis)
+					{
+						decltype(axis_placement->Axis)::type axis = axis_placement->Axis;
+						axis_direction = carve::geom::VECTOR(
+							axis->DirectionRatios[0],
+							axis->DirectionRatios[1],
+							axis->DirectionRatios[2]);
+					}
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: This was fine already - I assume that it got reverted?

**Review by pjanck — CHANGES_REQUESTED** [review]
Lots of stuff was good already but got reverted - making a good point for smaller commits ;)

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 26 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: I have used placementconverter::convertIfcDirection before, and it has caused exactly those changes to IFC files, that I was trying to revert. I guess I found a solution, how to correctly use placementconverter::convertIfcDirection and do not cause this problem anymore.

Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Found the solution. Resolved: 392efce

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 25 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 392efce

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 27 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 57e281f

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 24 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: db04c97

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 23 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: c3eef4a

**Comment by SamuilsRulovs** [comment]
I have refactored all functions except convertIfcSectionedSolidHorizontal (I guess it can be done as additional PR, since I do not know yet, if there should be any adjustments to this function) Right now SSH-Curve_1 looks like this: 
image

Other function which is not fully refactored is convertIfcHalfSpaceSolid, but it is completely commented out and, to my mind, it is unfinished, so I guess, it can stay like this, until it is finished.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 24 above)
Comment (pjanck) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Either it is:

- perpendicular to the axis -> it just stays as returned by the function
- fixed to the vertical direction -> the `,z` need correction

**Comment by pjanck** [comment]
Agreed, these functions can be noted as separate issues.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 28]
```diff
@@ -953,26 +955,9 @@ namespace OpenInfraPlatform
...
-					extrusionVector.x = vecDirection[0] * depth;
-					extrusionVector.y = vecDirection[1] * depth;
-					break;
-				}
-				default:
-					throw oip::InconsistentModellingException(extrudedArea, "Cardinality of the direction ratios can be only 2 or 3");
-				}
-				
+				carve::geom::vector<3> extrusionVector = placementConverter->convertIfcDirection(extrudedArea->ExtrudedDirection) 
+					* extrudedArea->Depth * UnitConvert()->getLengthInMeterFactor();
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: `this->UnitConvert()`

**Review by pjanck — CHANGES_REQUESTED** [review]

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 28 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 
5b15f2a

**Comment by pjanck** [comment]
The thing that I'm missing, is the alignment going through the solid in the example SSH_Curve_1 ... Should be noted as an issue.
